### ISSUE TITLE
Re-did my deployment by starting fresh, can't create user? (502 Bad Gateway)

### ISSUE THREAD

**Issue by GlassedSilver** [OpenYiff/Kemono2#170]
Not sure what I'm doing wrong. I started fresh for several reasons, so far, I'm getting Kemono served and everything seems to work okay, except for the issue described in the title. (well, haven't tested ingress of session keys so far either therefore, so who knows if I need help with one extra thing)

`docker ps` shows that it's constantly restarting???

nginx logs return this:

```
2022/04/03 07:16:08 [notice] 1#1: signal 3 (SIGQUIT) received, shutting down


2022/04/03 07:16:08 [notice] 31#31: gracefully shutting down


2022/04/03 07:16:08 [notice] 32#32: gracefully shutting down


2022/04/03 07:16:08 [notice] 31#31: exiting


2022/04/03 07:16:08 [notice] 37#37: gracefully shutting down


2022/04/03 07:16:08 [notice] 34#34: gracefully shutting down


2022/04/03 07:16:08 [notice] 38#38: gracefully shutting down


2022/04/03 07:16:08 [notice] 32#32: exiting


2022/04/03 07:16:08 [notice] 38#38: exiting


2022/04/03 07:16:08 [notice] 34#34: exiting


2022/04/03 07:16:08 [notice] 37#37: exiting


2022/04/03 07:16:08 [notice] 31#31: exit


2022/04/03 07:16:08 [notice] 32#32: exit


2022/04/03 07:16:08 [notice] 34#34: exit


2022/04/03 07:16:08 [notice] 38#38: exit


2022/04/03 07:16:08 [notice] 37#37: exit


2022/04/03 07:16:08 [notice] 35#35: gracefully shutting down


2022/04/03 07:16:08 [notice] 36#36: gracefully shutting down


2022/04/03 07:16:08 [notice] 36#36: exiting


2022/04/03 07:16:08 [notice] 36#36: exit


2022/04/03 07:16:08 [notice] 35#35: exiting


2022/04/03 07:16:08 [notice] 35#35: exit


2022/04/03 07:16:08 [notice] 33#33: gracefully shutting down


2022/04/03 07:16:08 [notice] 33#33: exiting


2022/04/03 07:16:08 [notice] 33#33: exit


2022/04/03 07:16:08 [notice] 1#1: signal 17 (SIGCHLD) received from 31


2022/04/03 07:16:08 [notice] 1#1: worker process 31 exited with code 0


2022/04/03 07:16:08 [notice] 1#1: worker process 37 exited with code 0


2022/04/03 07:16:08 [notice] 1#1: worker process 38 exited with code 0


2022/04/03 07:16:08 [notice] 1#1: signal 29 (SIGIO) received


2022/04/03 07:16:08 [notice] 1#1: signal 17 (SIGCHLD) received from 32


2022/04/03 07:16:08 [notice] 1#1: worker process 32 exited with code 0


2022/04/03 07:16:08 [notice] 1#1: worker process 34 exited with code 0


2022/04/03 07:16:08 [notice] 1#1: signal 29 (SIGIO) received


2022/04/03 07:16:08 [notice] 1#1: signal 17 (SIGCHLD) received from 35


2022/04/03 07:16:08 [notice] 1#1: worker process 35 exited with code 0


2022/04/03 07:16:08 [notice] 1#1: worker process 36 exited with code 0


2022/04/03 07:16:08 [notice] 1#1: signal 29 (SIGIO) received


2022/04/03 07:16:08 [notice] 1#1: signal 17 (SIGCHLD) received from 36


2022/04/03 07:16:08 [notice] 1#1: signal 17 (SIGCHLD) received from 33


2022/04/03 07:16:08 [notice] 1#1: worker process 33 exited with code 0


2022/04/03 07:16:08 [notice] 1#1: exit


/docker-entrypoint.sh: /docker-entrypoint.d/ is not empty, will attempt to perform configuration


/docker-entrypoint.sh: Looking for shell scripts in /docker-entrypoint.d/


/docker-entrypoint.sh: Launching /docker-entrypoint.d/10-listen-on-ipv6-by-default.sh


10-listen-on-ipv6-by-default.sh: info: IPv6 listen already enabled


/docker-entrypoint.sh: Launching /docker-entrypoint.d/20-envsubst-on-templates.sh


/docker-entrypoint.sh: Launching /docker-entrypoint.d/30-tune-worker-processes.sh


/docker-entrypoint.sh: Configuration complete; ready for start up


2022/04/03 07:16:32 [notice] 1#1: using the "epoll" event method


2022/04/03 07:16:32 [notice] 1#1: nginx/1.21.6


2022/04/03 07:16:32 [notice] 1#1: built by gcc 10.2.1 20210110 (Debian 10.2.1-6) 


2022/04/03 07:16:32 [notice] 1#1: OS: Linux 5.4.0-107-generic


2022/04/03 07:16:32 [notice] 1#1: getrlimit(RLIMIT_NOFILE): 1048576:1048576


2022/04/03 07:16:32 [notice] 1#1: start worker processes


2022/04/03 07:16:32 [notice] 1#1: start worker process 24


2022/04/03 07:16:32 [notice] 1#1: start worker process 25


2022/04/03 07:16:32 [notice] 1#1: start worker process 26


2022/04/03 07:16:32 [notice] 1#1: start worker process 27


2022/04/03 07:16:32 [notice] 1#1: start worker process 28


2022/04/03 07:16:32 [notice] 1#1: start worker process 29


2022/04/03 07:16:32 [notice] 1#1: start worker process 30


2022/04/03 07:16:32 [notice] 1#1: start worker process 31


2022/04/03 07:17:24 [warn] 24#24: *1 an upstream response is buffered to a temporary file /var/cache/nginx/proxy_temp/1/00/0000000001 while reading upstream, client: 192.168.2.220, server: , request: "GET /static/bundle/js/global.bundle.js HTTP/1.1", upstream: "http://172.20.0.6:80/static/bundle/js/global.bundle.js", host: "192.168.2.183:5000", referrer: "http://192.168.2.183:5000/"


2022/04/03 07:19:29 [error] 28#28: *14 upstream prematurely closed connection while reading response header from upstream, client: 192.168.2.220, server: , request: "POST /account/login HTTP/1.1", upstream: "http://172.20.0.6:80/account/login", host: "192.168.2.183:5000", referrer: "http://192.168.2.183:5000/account/login"


2022/04/03 07:19:35 [error] 29#29: *11 upstream prematurely closed connection while reading response header from upstream, client: 192.168.2.220, server: , request: "POST /account/login HTTP/1.1", upstream: "http://172.20.0.6:80/account/login", host: "192.168.2.183:5000", referrer: "http://192.168.2.183:5000/account/login"


2022/04/03 07:22:10 [warn] 31#31: *36 an upstream response is buffered to a temporary file /var/cache/nginx/proxy_temp/2/00/0000000002 while reading upstream, client: 192.168.2.220, server: , request: "GET /static/bundle/js/global.bundle.js HTTP/1.1", upstream: "http://172.20.0.6:80/static/bundle/js/global.bundle.js", host: "192.168.2.183:5000", referrer: "http://192.168.2.183:5000/account/login"


2022/04/03 07:23:03 [error] 31#31: *36 upstream prematurely closed connection while reading response header from upstream, client: 192.168.2.220, server: , request: "POST /account/register HTTP/1.1", upstream: "http://172.20.0.6:80/account/register", host: "192.168.2.183:5000", referrer: "http://192.168.2.183:5000/account/register"


2022/04/03 07:23:15 [warn] 30#30: *33 an upstream response is buffered to a temporary file /var/cache/nginx/proxy_temp/3/00/0000000003 while reading upstream, client: 192.168.2.220, server: , request: "GET /static/bundle/js/global.bundle.js HTTP/1.1", upstream: "http://172.20.0.6:80/static/bundle/js/global.bundle.js", host: "192.168.2.183:5000", referrer: "http://192.168.2.183:5000/account/login"


2022/04/03 07:23:20 [warn] 30#30: *33 an upstream response is buffered to a temporary file /var/cache/nginx/proxy_temp/4/00/0000000004 while reading upstream, client: 192.168.2.220, server: , request: "GET /static/bundle/js/global.bundle.js HTTP/1.1", upstream: "http://172.20.0.6:80/static/bundle/js/global.bundle.js", host: "192.168.2.183:5000", referrer: "http://192.168.2.183:5000/importer"


2022/04/03 07:23:33 [error] 30#30: *33 upstream prematurely closed connection while reading response header from upstream, client: 192.168.2.220, server: , request: "POST /account/register HTTP/1.1", upstream: "http://172.20.0.6:80/account/register", host: "192.168.2.183:5000", referrer: "http://192.168.2.183:5000/account/register"


2022/04/03 07:24:45 [error] 30#30: *33 upstream prematurely closed connection while reading response header from upstream, client: 192.168.2.220, server: , request: "POST /account/register HTTP/1.1", upstream: "http://172.20.0.6:80/account/register", host: "192.168.2.183:5000", referrer: "http://192.168.2.183:5000/account/register"


2022/04/03 07:26:51 [error] 26#26: *81 upstream prematurely closed connection while reading response header from upstream, client: 192.168.2.220, server: , request: "POST /account/register HTTP/1.1", upstream: "http://172.20.0.6:80/account/register", host: "192.168.2.183:5000", referrer: "http://192.168.2.183:5000/account/register"
```

**Comment by BassOfBass** [comment]
For prod config deployment you have to dive into the containers to read logs.
So first off you have to run `docker ps --all` and tell the server names which restart.

**Comment by GlassedSilver** [comment]
```
CONTAINER ID   IMAGE                           COMMAND                  CREATED        STATUS                          PORTS                                                                                            NAMES
375829a2b726   portainer/portainer-ce:latest   "/portainer"             13 hours ago   Up 13 hours                     0.0.0.0:8000->8000/tcp, :::8000->8000/tcp, 0.0.0.0:9443->9443/tcp, :::9443->9443/tcp, 9000/tcp   portainer
362d078a5c9d   nginx                           "/docker-entrypoint.…"   14 hours ago   Up 9 hours                      0.0.0.0:5000->80/tcp, :::5000->80/tcp                                                            kemono-nginx
060d1ce526b3   kemono-2_kemono-archiver        "/bin/sh -c 'python3…"   14 hours ago   Restarting (1) 27 seconds ago                                                                                                    kemono-archiver
eb72d126ad23   kemono-2_kemono-app-dev         "/bin/sh -c 'uwsgi -…"   14 hours ago   Up 14 hours (healthy)                                                                                                            kemono
ac0f13ef9ea2   redis:6.2.0                     "docker-entrypoint.s…"   14 hours ago   Up 14 hours                     6379/tcp                                                                                         kemono-redis
be24929ccb7b   healthcheck/postgres            "docker-entrypoint.s…"   14 hours ago   Up 14 hours (healthy)           5432/tcp                                                                                         kemono-db
6f7df8c096ac   kemono-2_kemono-frontend-dev    "docker-entrypoint.s…"   14 hours ago   Up 14 hours                                                                                                                      kemono-frontend-dev
8b62e95a453f   willfarrell/autoheal            "/docker-entrypoint …"   14 hours ago   Up 14 hours (healthy)                                                                                                            kemono-autoheal
```

Forgot to mention that it is nginx that is constantly restarting, but yeah this would be `docker ps —all` rather than just `docker ps`.

And that log I posted is in fact the log for kemono‘s docker compose included nginx container. Any other container that might be useful?

I have a suspicion that nginx writing some stuff to a temporary file is not what it should be doing.

Probably shouldn‘t be using the dev compose file, right? I only went with that because the instructions explained to use it, but then again the instructions focus on running a dev environment to begin with. (which is understandable, since most folks are probably making use of this to work on kemono‘s code and UI)

**Comment by GlassedSilver** [comment]
Okay I‘m using the non-dev docker compose file now, but after a successful build I‘m running into archiver logging this error repeatedly:

```
Traceback (most recent call last):
  File "daemon.py", line 21, in <module>
    backend = get_backend(pg_url)
  File "/usr/local/lib/python3.8/site-packages/yoyo/connections.py", line 99, in get_backend
    backend = backend_class(parsed, migration_table)
  File "/usr/local/lib/python3.8/site-packages/yoyo/backends.py", line 165, in __init__
    self._connection = self.connect(dburi)
  File "/usr/local/lib/python3.8/site-packages/yoyo/backends.py", line 706, in connect
    return self.driver.connect(**kwargs)
  File "/usr/local/lib/python3.8/site-packages/psycopg2/__init__.py", line 127, in connect
    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
psycopg2.OperationalError: FATAL:  password authentication failed for user "nano“
```

**Comment by BassOfBass** [comment]
> I have a suspicion that nginx writing some stuff to a temporary file is not what it should be doing.

Nginx doesn't do anything wrong, it just won't start until all other containers in the stack properly start.
It's `kemono-archiver` crashing, probably because of `development/assets` folder.
> I only went with that because the instructions explained to use it, but then again the instructions focus on running a dev environment to begin with.

The dev config is for development. There are no instructions for prod because the prod for site is using a quite different setup.

**Comment by BassOfBass** [comment]
> Okay I‘m using the non-dev docker compose file now, but after a successful build I‘m running into archiver logging this error repeatedly:

Probably because you have different passwords between archiver and server folders.

**Comment by GlassedSilver** [comment]
Okay, so the things I configured are as follows, maybe you can tell me what I may have missed. :)

`.env` file:
```
PGHOST=localhost
PGUSER=nano
PGPASSWORD="MYPASSWORD-REDACTED"
PGDATABASE=kemonodb

TELEGRAMTOKEN=
TELEGRAMCHANNEL=

REDIS_HOST=kemono-redis
REDIS_PORT=6379

KEMONO_SITE=http://localhost:5000
# ANTIANTISCRAPER_PROXY=
```

`kitsune.py` file:
```
download_path = '/storage'

# database info
database_host = 'kemono-db'
database_dbname = 'kemonodb'
database_user = 'nano'
database_password = 'MYPASSWORD-REDACTED'
database_port = 5432

redis_host = 'kemono-redis'
redis_port = 6379

# proxies = ["socks5://user:pass@host:port"]
proxies = []

# BAN url prefix (for varnish purging and the like)
# ban_url = 'http://10.0.0.1:8313'
ban_url = ''

# public encryption key for key contribution and autoimport
# generate one with `openssl genrsa -out privatekey.txt 4096 && openssl rsa -in privatekey.txt -pubout -out publickey && cat publickey`
# as seen above, it is highly recommended you use 4096bit keys... session tokens can get too long for 2048 quite easily.
pubkey = """
mypubkey-redacted (generated into ~/kemono-2)
"""

# (static) salt for autoimport key hashing. you should change this once and keep it safe.
salt = "MYSALT-REDACTED"

# enable/disable redis key pushing
pubsub = True
pubsub_queue_limit = 200
```

`redis_map.py` is unchanged because my redis is password-less

`flask.cfg` file:
```
CACHE_TYPE = 'simple'
CACHE_DEFAULT_TIMEOUT = 60
ENABLE_PASSWORD_VALIDATOR=True
ENABLE_LOGIN_RATE_LIMITING=True
SECRET_KEY = '~/kemono-2/privatekey.txt'
```

I also copied `kitsune.py`'s contents into `~/kemono-2/archiver/config.py` & I copied `redis_map.py` into that same sub-directory. (basically I followed [your instructions here](https://github.com/OpenYiff/Kemono2/issues/150#issuecomment-1002917979))

Then, being in `~/kemono-2/` I executed

1. `docker compose --file docker-compose.dev.yml build` (build is successful)
2. `docker compose --file docker-compose.dev.yml up` (errors as above)

**Comment by BassOfBass** [comment]
I have no idea what is the problem, try to use the default password first.
Maybe you are running the prod db on the preexisting dev db volume.

**Comment by GlassedSilver** [comment]
I fixed it. Turns out my docker compose file still had the wrong password. Ooops.

I also set up binding the volumes to a host path, the beauty of that is that I can store the assets on my hypervisor that hosts the docker compose VM anywhere I want now. So all those big files don't need to sit on my SSD where my VMs live, but media files get stored on my HDD array. Really sweet.

Created an account and that works. Ran into another issue now though, so I'll close this one and post it in a new one for structure. :)

Thank you so much for your help and time, it's very appreciated! I think I'm getting a lot closer to finally having a very solid setup! :)
